Tool calls missing arguments field causing AI_TypeValidationError (anthropic haiku-4.5)

When using Anthropic Claude Haiku through OpenRouter, tool calls are returned without the required `arguments` field, causing AI SDK v5 validation to fail.

**Error:**
`AI_TypeValidationError: Invalid input: expected string, received undefined at path ["choices",0,"message","tool_calls",0,"function","arguments"]`

**Response from OpenRouter:**
{
  "tool_calls": [{
    "id": "toolu_bdrk_...",
    "type": "function",
    "function": {
      "name": "list_available_components"
      // Missing: "arguments" field
    }
  }]
}

**Expected:** `"arguments": "{}"` or `"arguments": ""`

**Version:** @openrouter/ai-sdk-provider@1.2.8, ai@^5.0.0


```
AI_APICallError: Invalid JSON response
 ❯ ../../node_modules/@openrouter/ai-sdk-provider/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.1_zod@3.25.76/node_modules/@ai-sdk/provider-utils/src/response-handler.ts:151:13
 ❯ postToApi ../../node_modules/@openrouter/ai-sdk-provider/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.1_zod@3.25.76/node_modules/@ai-sdk/provider-utils/src/post-to-api.ts:136:14
 ❯ OpenRouterChatLanguageModel.doGenerate ../../node_modules/@openrouter/ai-sdk-provider/src/chat/index.ts:226:55
 ❯ fn node_modules/ai/src/generate-text/generate-text.ts:389:32
 ❯ node_modules/ai/src/telemetry/record-span.ts:18:22
 ❯ _retryWithExponentialBackoff node_modules/ai/src/util/retry-with-exponential-backoff.ts:96:12
 ❯ fn node_modules/ai/src/generate-text/generate-text.ts:345:34
 ❯ node_modules/ai/src/telemetry/record-span.ts:18:22
 ❯ Module.generateText node_modules/ai/src/generate-text/generate-text.ts:281:12

Caused by: AI_TypeValidationError: Type validation failed: Value: {"id":"gen-1765191297-cCNRPMadSsGK3fCO5kmn","provider":"Amazon Bedrock","model":"anthropic/claude-haiku-4.5","object":"chat.completion","created":1765191297,"choices":[{"logprobs":null,"finish_reason":"tool_calls","native_finish_reason":"tool_calls","index":0,"message":{"role":"assistant","content":"","refusal":null,"reasoning":null,"tool_calls":[{"id":"toolu_bdrk_015kDTLLfWrSmkqgiahrdeCV","index":0,"type":"function","function":{"name":"list_available_components"}}]}}],"usage":{"prompt_tokens":5435,"completion_tokens":39,"total_tokens":5474,"cost":0.0010382,"is_byok":false,"prompt_tokens_details":{"cached_tokens":5102,"audio_tokens":0,"video_tokens":0},"cost_details":{"upstream_inference_cost":null,"upstream_inference_prompt_cost":0.0008432,"upstream_inference_completions_cost":0.000195},"completion_tokens_details":{"reasoning_tokens":0,"image_tokens":0}}}.
Error message: [{"code":"invalid_union","errors":[[{"expected":"string","code":"invalid_type","path":["choices",0,"message","tool_calls",0,"function","arguments"],"message":"Invalid input: expected string, received undefined"}],[{"expected":"object","code":"invalid_type","path":["error"],"message":"Invalid input: expected object, received undefined"}]],"path":[],"message":"Invalid input"}]
 ❯ Object.validate ../../node_modules/@openrouter/ai-sdk-provider/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.1_zod@3.25.76/node_modules/@ai-sdk/provider-utils/src/validator.ts:69:18
 ❯ safeValidateTypes ../../node_modules/@openrouter/ai-sdk-provider/node_modules/.pnpm/@ai-sdk+provider-utils@3.0.1_zod@3.25.76/node_modules/@ai-sdk/provider-utils/src/validate-types.ts:64:20
```

--- a/.changeset/fix-tool-calls-missing-arguments.md
+++ b/.changeset/fix-tool-calls-missing-arguments.md
@@ -1,0 +1,7 @@
+---
+"@openrouter/ai-sdk-provider": patch
+---
+
+fix: handle tool calls with missing arguments field (#287)
+
+Made the arguments field optional in the tool_calls schema and default to '{}' (empty JSON object) when missing. This handles cases where upstream providers may omit the arguments field for tools with no parameters.

--- a/src/chat/index.test.ts
+++ b/src/chat/index.test.ts
@@ -529,7 +529,7 @@ describe('doGenerate', () => {
     );
   });
 
-  it('should handle tool calls with missing arguments field (Anthropic Haiku)', async () => {
+  it('should default to empty JSON object when tool call arguments field is missing', async () => {
     prepareJsonResponse({
       content: '',
       tool_calls: [
@@ -538,7 +538,6 @@ describe('doGenerate', () => {
           type: 'function',
           function: {
             name: 'get_current_time',
-            // arguments field is omitted - some models like Anthropic Haiku do this
           },
         },
       ],
@@ -549,7 +548,6 @@ describe('doGenerate', () => {
       prompt: TEST_PROMPT,
     });
 
-    // Should default to empty JSON object when arguments is missing
     expect(result.content).toContainEqual(
       expect.objectContaining({
         type: 'tool-call',

--- a/src/chat/index.ts
+++ b/src/chat/index.ts
@@ -390,7 +390,6 @@ export class OpenRouterChatLanguageModel implements LanguageModelV3 {
           type: 'tool-call' as const,
           toolCallId: toolCall.id ?? generateId(),
           toolName: toolCall.function.name,
-          // Some models (e.g., Anthropic Haiku) may omit arguments field when there are no arguments
           input: toolCall.function.arguments ?? '{}',
           providerMetadata: !reasoningDetailsAttachedToToolCall
             ? {

--- a/src/chat/schemas.ts
+++ b/src/chat/schemas.ts
@@ -62,8 +62,6 @@ export const OpenRouterNonStreamChatCompletionResponseSchema = z.union([
                       function: z
                         .object({
                           name: z.string(),
-                          // Some models (e.g., Anthropic Haiku) may omit arguments field
-                          // when there are no arguments to pass
                           arguments: z.string().optional(),
                         })
                         .passthrough(),

